pkg/scheduler/apis/config/v1alpha2/conversion.go

```diff
@@ -0,0 +1,67 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha2
+
+import (
+	conversion "k8s.io/apimachinery/pkg/conversion"
+	"k8s.io/component-base/config/v1alpha1"
+	v1alpha2 "k8s.io/kube-scheduler/config/v1alpha2"
+	"k8s.io/kubernetes/pkg/scheduler/apis/config"
+	"unsafe"
+)
+
+func Convert_v1alpha2_KubeSchedulerConfiguration_To_config_KubeSchedulerConfiguration(in *v1alpha2.KubeSchedulerConfiguration, out *config.KubeSchedulerConfiguration, s conversion.Scope) error {
+	if err := autoConvert_v1alpha2_KubeSchedulerConfiguration_To_config_KubeSchedulerConfiguration(in, out, s); err != nil {
+		return err
+	}
+	val := v1alpha2.SchedulerDefaultProviderName
+	out.AlgorithmSource.Provider = &val
+	return nil
+}
+
+func Convert_config_KubeSchedulerConfiguration_To_v1alpha2_KubeSchedulerConfiguration(in *config.KubeSchedulerConfiguration, out *v1alpha2.KubeSchedulerConfiguration, s conversion.Scope) error {
+	out.TypeMeta = in.TypeMeta
+	out.SchedulerName = &in.SchedulerName
+	out.HardPodAffinitySymmetricWeight = &in.HardPodAffinitySymmetricWeight
+	if err := Convert_config_KubeSchedulerLeaderElectionConfiguration_To_v1alpha2_KubeSchedulerLeaderElectionConfiguration(&in.LeaderElection, &out.LeaderElection, s); err != nil {
+		return err
+	}
+	if err := v1alpha1.Convert_config_ClientConnectionConfiguration_To_v1alpha1_ClientConnectionConfiguration(&in.ClientConnection, &out.ClientConnection, s); err != nil {
+		return err
+	}
+	out.HealthzBindAddress = &in.HealthzBindAddress
+	out.MetricsBindAddress = &in.MetricsBindAddress
+	if err := v1alpha1.Convert_config_DebuggingConfiguration_To_v1alpha1_DebuggingConfiguration(&in.DebuggingConfiguration, &out.DebuggingConfiguration, s); err != nil {
+		return err
+	}
+	out.DisablePreemption = &in.DisablePreemption
+	out.PercentageOfNodesToScore = &in.PercentageOfNodesToScore
+	out.BindTimeoutSeconds = &in.BindTimeoutSeconds
+	out.PodInitialBackoffSeconds = &in.PodInitialBackoffSeconds
+	out.PodMaxBackoffSeconds = &in.PodInitialBackoffSeconds
+	if in.Plugins != nil {
+		in, out := &in.Plugins, &out.Plugins
+		*out = new(v1alpha2.Plugins)
+		if err := Convert_config_Plugins_To_v1alpha2_Plugins(*in, *out, s); err != nil {
+			return err
+		}
+	} else {
+		out.Plugins = nil
+	}
+	out.PluginConfig = *(*[]v1alpha2.PluginConfig)(unsafe.Pointer(&in.PluginConfig))
+	return nil
```

Comment:
This one is tricky, but I'm not sure if we should care.

Theoretically, if there is an AlgorithmSource in the internal configuration, we should obtain the diff that needs to be applied to Plugins and Pluginconfig. But this is only needed if someone tries to do something like v1alpha1->internal->v1alpha2.

This is something that can be done by passing `--config` in v1alpha1 and `--write-config-to`, which writes to v1alpha2 all the time.

Again, should we care? I would say it's not worth the effort.

Reply:
According to the API conventions, we need to be able to round-trip to older versions without losing any information (see item 4 in this list: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#on-compatibility)

So that brings up a good question, because like you suggested [here](https://github.com/kubernetes/kubernetes/pull/87999#discussion_r377812933) we can just use the autoConvert function, but how do we handle roundtripping AlgorithmSource? We need to keep that info somewhere (like an annotation?), or we can just drop it once the Scheduler no longer supports AlgorithmSource.

And in this point, how can we get the information from AlgorithmSource that needs to be applied to Plugins and PluginConfig? If it's a configmap, I suppose we could read from that, but if it's a policy file that becomes tougher.

Reply:
Better information more specific to deprecation here, but I'm still not clear. I think an annotation is the way it suggests? https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-parts-of-the-api

Reply:
Though, this api is only in alpha so it might not be relevant for us. I'm ok with just dropping the information too

Reply:
I would expect the round-trip requirement to be specific to APIs in which multiple clients interact with different versions of the same resource (e.g. REST APIs).

API types used in config files are input-only, and have multiple producers but a single consumer (the component whose configuration they are driving)

Reply:
I think our policy doesn't cover it while alpha either way.

I think it's confusing at best to have config file formats that are like our api objects but have different rules. I don't have specific case in which they go wrong, though. But I expect the inconsistency will eventually be a problem, e.g. maybe people start making these into CRDs for some strange reason. Anyway, I don't have concerns for this particular change.

Reply:
In general, these configuration types are not supposed to be outputted. If we could add some type of annotation in them such that no internal->versioned conversion code is auto-generated, that would be a win already.